### DISCUSSION TITLE
Monitoring VIPs on client interfaces

### DISCUSSION THREAD

**Discussion by ventureElen** [strongswan/strongswan#2452] [Q&A] [answered]
**Problem statement:**
we have encountered an issue in which Client is using same interface (let say eth0) for both outer and inner IPs.
Outer IP is given by DHCP server.
Inner IP is VIP. it will be installed during IPsec tunnel creation phase. ( Server will install a random IP from pool).

Occasionally during heavy load,  DHCP server fails to renew the outer IP. as a result, dhcp client releases the IP and requests new IP from DHCP server. DHCP server gives same IP in response. **outer IP gets restored but VIP is deleted during this process.**
As soon as client regains the IP (within few sec), it gets stuck in DPD loop. it keeps on sending DPD message to the server every 30 sec.

**Queries:**
1. Does strongswan or any of its plugins support monitoring of VIP on client interfaces (to automatically detect and plumb VIP if it is missing or deleted)
2. Can we configure vpn client/server to terminate & recreate the tunnel if client is stuck in infinite loop in which it sends DPD message every 30 sec?

**Client logs:**
`9419:Sep 17 06:37:16 info 15[IKE] sending DPD request
9428:Sep 17 06:37:46 info 09[IKE] sending DPD request
9437:Sep 17 06:38:16 info 15[IKE] sending DPD request
9444:Sep 17 06:38:46 info 10[IKE] sending DPD request
9451:Sep 17 06:39:16 info 11[IKE] sending DPD request
9458:Sep 17 06:39:46 info 08[IKE] sending DPD request
9465:Sep 17 06:40:16 info 05[IKE] sending DPD request
9472:Sep 17 06:40:46 info 13[IKE] sending DPD request
9479:Sep 17 06:41:16 info 10[IKE] sending DPD request
9490:Sep 17 06:41:46 info 07[IKE] sending DPD request
9497:Sep 17 06:42:16 info 08[IKE] sending DPD request
9502:Sep 17 06:42:46 info 12[IKE] sending DPD request
9509:Sep 17 06:43:16 info 14[IKE] sending DPD request
9514:Sep 17 06:43:46 info 07[IKE] sending DPD request
9526:Sep 17 06:44:16 info 08[IKE] sending DPD request
9531:Sep 17 06:44:46 info 12[IKE] sending DPD request
9538:Sep 17 06:45:16 info 14[IKE] sending DPD request
9543:Sep 17 06:45:46 info 07[IKE] sending DPD request
9550:Sep 17 06:46:16 info 08[IKE] sending DPD request
9567:Sep 17 06:46:46 info 12[IKE] sending DPD request
9579:Sep 17 06:47:16 info 14[IKE] sending DPD request
9592:Sep 17 06:47:46 info 07[IKE] sending DPD request
9599:Sep 17 06:48:16 info 08[IKE] sending DPD request
9604:Sep 17 06:48:46 info 12[IKE] sending DPD request
9611:Sep 17 06:49:16 info 14[IKE] sending DPD request
9616:Sep 17 06:49:46 info 12[IKE] sending DPD request
9623:Sep 17 06:50:16 info 14[IKE] sending DPD request
9633:Sep 17 06:50:46 info 07[IKE] sending DPD request
9640:Sep 17 06:51:16 info 08[IKE] sending DPD request
9645:Sep 17 06:51:46 info 12[IKE] sending DPD request
9652:Sep 17 06:52:16 info 14[IKE] sending DPD request
9663:Sep 17 06:52:46 info 07[IKE] sending DPD request
9670:Sep 17 06:53:16 info 08[IKE] sending DPD request
9675:Sep 17 06:53:46 info 12[IKE] sending DPD request
9682:Sep 17 06:54:16 info 14[IKE] sending DPD request
9687:Sep 17 06:54:46 info 07[IKE] sending DPD request
9694:Sep 17 06:55:16 info 08[IKE] sending DPD request
9699:Sep 17 06:55:46 info 12[IKE] sending DPD request
9706:Sep 17 06:56:16 info 14[IKE] sending DPD request
9711:Sep 17 06:56:46 info 07[IKE] sending DPD request
9718:Sep 17 06:57:16 info 08[IKE] sending DPD request
9723:Sep 17 06:57:46 info 12[IKE] sending DPD request
9730:Sep 17 06:58:16 info 14[IKE] sending DPD request
9735:Sep 17 06:58:46 info 07[IKE] sending DPD request
9742:Sep 17 06:59:16 info 14[IKE] sending DPD request
9747:Sep 17 06:59:46 info 07[IKE] sending DPD request
9755:Sep 17 07:00:16 info 08[IKE] sending DPD request
9760:Sep 17 07:00:46 info 12[IKE] sending DPD request
9767:Sep 17 07:01:16 info 14[IKE] sending DPD request
9772:Sep 17 07:01:46 info 07[IKE] sending DPD request
9779:Sep 17 07:02:16 info 08[IKE] sending DPD request
9784:Sep 17 07:02:46 info 12[IKE] sending DPD request
9791:Sep 17 07:03:16 info 14[IKE] sending DPD request
9796:Sep 17 07:03:46 info 07[IKE] sending DPD request
9803:Sep 17 07:04:16 info 08[IKE] sending DPD request
9808:Sep 17 07:04:46 info 12[IKE] sending DPD request
9815:Sep 17 07:05:16 info 14[IKE] sending DPD request
9820:Sep 17 07:05:46 info 07[IKE] sending DPD request
9827:Sep 17 07:06:16 info 08[IKE] sending DPD request
9832:Sep 17 07:06:46 info 12[IKE] sending DPD request
9839:Sep 17 07:07:16 info 14[IKE] sending DPD request
9846:Sep 17 07:07:46 info 07[IKE] sending DPD request
9853:Sep 17 07:08:16 info 14[IKE] sending DPD request
9858:Sep 17 07:08:46 info 07[IKE] sending DPD request
9865:Sep 17 07:09:16 info 08[IKE] sending DPD request
9870:Sep 17 07:09:46 info 12[IKE] sending DPD request
9878:Sep 17 07:10:16 info 14[IKE] sending DPD request
9883:Sep 17 07:10:46 info 07[IKE] sending DPD request
9890:Sep 17 07:11:16 info 08[IKE] sending DPD request
9895:Sep 17 07:11:46 info 12[IKE] sending DPD request
9902:Sep 17 07:12:16 info 14[IKE] sending DPD request
9907:Sep 17 07:12:46 info 07[IKE] sending DPD request
9914:Sep 17 07:13:16 info 08[IKE] sending DPD request
9919:Sep 17 07:13:46 info 12[IKE] sending DPD request
9926:Sep 17 07:14:16 info 14[IKE] sending DPD request
9931:Sep 17 07:14:46 info 07[IKE] sending DPD request
9938:Sep 17 07:15:16 info 08[IKE] sending DPD request
9943:Sep 17 07:15:46 info 12[IKE] sending DPD request
9950:Sep 17 07:16:16 info 14[IKE] sending DPD request`

**Comment by tobiasbrunner** [comment, marked as answer]
> 1. Does strongswan or any of its plugins support monitoring of VIP on client interfaces (to automatically detect and plumb VIP if it is missing or deleted)

VIPs are reinstalled if the CHILD_SA is updated (i.e. its outer addresses change). But if the IP address stays the same, there won't be any update. So I suggest installing the VIP on a stable interface like `lo` so it doesn't disappear and the required route can be reinstalled when the dynamic interface comes up again.

> 2\. Can we configure vpn client/server to terminate & recreate the tunnel if client is stuck in infinite loop in which it sends DPD message every 30 sec?

There is not really any loop. DPDs are simply exchanged successfully.  But DPDs really only check the liveness of the IKE_SA, they don'tt say anything about whether traffic can be exchanged via CHILD_SAs. While the fact that a DPD is sent in the first place means there was no traffic from the peer within the configured time (IKE or ESP), that's not an error condition. To verify whether traffic can be exchanged, you'd have to do your own checks (e.g. ping a remote host). What you could configure is an inactivity timeout. However, that will just close the CHILD_SA after there was no traffic for a while, it won't get recreated unless you use trap policies and outbound traffic occurs or you initiate the SA again yourself.
